### PR TITLE
ParseXS: extensive refactoring of param processing, and more tests

### PR DESCRIPTION
**Pull request by iabyn**
This set of about 33 commits:
- improves the testing infrastructure of ExtUtils::ParseXS - in particular, it makes it possible to include short XS 'files'  as heredocs within foo.t files, rather than having to create a separate t/XSfoo.xs file each time;
- adds lots of new tests;
- adds some new error conditions;
- heavily refactors the code which deals with the parsing and code generation of XSUB parameters,
- and in particular, stops ANSI parameters being handled by the injection of a fake INPUT section.
There's still more work to do on cleaning up parameter processing, but this was a good place to pause.

### CODE REVIEW

File: dist/ExtUtils-ParseXS/t/001-basic.t [hunk 1]
```diff
@@ -545,3 +545,40 @@ EOF
 
     like($sout, qr/pkg.*=.*"Foo"/, "default expression expanded");
 }
+
+{
+    # Test 'alien' INPUT parameters: ones which are declared in an INPUT
+    # section but don't appear in the XSUB's signature. This aught to be
```
Comment (JRaspass) [inline dist/ExtUtils-ParseXS/t/001-basic.t]: ought, not aught, surely? Also once more in this para and twice in the commit msg.

File: dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm [hunk 2]
```diff
@@ -583,8 +595,13 @@ EOM
 
   # Open the input file (using $self->{in_filename} which
   # is a basename'd $Options{filename} due to chdir above)
-  open($self->{in_fh}, '<', $self->{in_filename})
-      or die "cannot open $self->{in_filename}: $!\n";
+  {
+    my $fn   = $self->{in_filename};
+    my $opfn = $Options{filename};
+    $fn = $opfn if ref $opfn; # allow string ref as a file handle
```
Comment (tonycoz) [inline dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm]: `$opfn` (or `$$opfn`) is the source of the file, not a file handle

File: dist/ExtUtils-ParseXS/t/001-basic.t [hunk 3]
```diff
@@ -582,3 +582,30 @@ EOF
 
     like($sout, qr/longalien1;\nintalien2=123;/, "alien INPUT parameters");
 }
+
+{
+    # Test for 'No INPUT definition' error, particularly that the
+    # type is outout correctly in the error message.
```
Comment (tonycoz) [inline dist/ExtUtils-ParseXS/t/001-basic.t]: outout -> output?

**Comment by tonycoz** [comment]
Looks fine except for the minor issues that have been commented on.

**Pull request merged**